src/Hamiltonians/HOCartesian2BosonRelative.jl

```diff
@@ -0,0 +1,230 @@
+        interaction_only = false,
+        warn_address = true
+    )
+    if isnothing(ηs) && isnothing(Sx)
+        S[1] ≠ maximum(S) && throw(ArgumentError("First dimension must have the most states"))
+        aspect = float.(box_to_aspect(S))
+        D = length(S)
+    else
+        any(ηs .< 1.0) && throw(ArgumentError("Aspect ratios must be greater than 1.0"))
+        D = length(ηs) + 1
+        Srest = map(x -> floor(Int, (Sx - 1) / x + 1), ηs)
+        S = (Sx, Srest...)
+        aspect = (1.0, float.(ηs)...)
+    end
+    P = prod(S)
+
+    # address type is fixed so it is overwritten here to ensure it works
+    warn_address && !(addr isa BoseFS{1,P}) && @warn "Input address is being overwritten"
+    addr = BoseFS(P, 1 => 1)
+
+    M = maximum(S) - 1
+    v_vec = [log_abs_oscillator_zero(i) for i in 0:2:M]
+
+    return HOCartesian2BosonRelative{D,typeof(addr)}(addr, S, aspect, v_vec, g, interaction_only)
+end
+
+function Base.show(io::IO, H::HOCartesian2BosonRelative)
+    if length(H.S) == 1
+        print(io, "HOCartesian2BosonRelative($(H.addr); S=$(H.S), g=$(H.g), interaction_only=$(H.interaction_only))")
+    else
+        print(io, "HOCartesian2BosonRelative($(H.addr); Sx=$(H.S[1]), ηs=$(H.aspect[2:end]), g=$(H.g), interaction_only=$(H.interaction_only))")
+    end
+end
+
+Base.:(==)(H::HOCartesian2BosonRelative, G::HOCartesian2BosonRelative) = all(map(p -> getproperty(H, p) == getproperty(G, p), propertynames(H)))
+
+starting_address(H::HOCartesian2BosonRelative) = H.addr
+
+LOStructure(::Type{<:HOCartesian2BosonRelative}) = IsHermitian()
+
+### DIAGONAL ELEMENTS ###
+@inline function noninteracting_energy(S, aspect, omm::BoseOccupiedModeMap)
+    states = CartesianIndices(S)
+    return sum(omm) do p
+        indices = Tuple(states[p.mode]) .- 1
+        dot(indices, aspect) + sum(aspect)/2
+    end
+end
+noninteracting_energy(H::HOCartesian2BosonRelative, addr) = noninteracting_energy(H.S, H.aspect, OccupiedModeMap(addr))
+
+@inline function diagonal_element(H::HOCartesian2BosonRelative, addr)
+    omm = OccupiedModeMap(addr)
+    u = H.g / sqrt(prod(H.aspect))
+    result = u * ho_2brel_interaction(H.S, H.vtable, omm, omm)
+    if !H.interaction_only
+        result += noninteracting_energy(H, addr)
+    end
+    return result
+end
+
+### OFFDIAGONAL ELEMENTS ###
+# all addresses couple
+num_offdiagonals(H::HOCartesian2BosonRelative, addr) = prod(H.S) - 1
+
+function get_offdiagonal(
+        H::HOCartesian2BosonRelative, 
+        addr, 
+        chosen::Int, 
+        omm::OccupiedModeMap = OccupiedModeMap(addr)
+    )
+    P = prod(H.S)
+    chosen ≥ P - 1 && return addr, 0.0
+
+    # skip self-move
+    p_i = omm[1]
+    if chosen ≥ p_i.mode
+        chosen += 2     # conserve parity
+    end
+
+    p_j = find_mode(addr, chosen)
+    new_addr, val = excitation(addr, (p_j,), (p_i,))
+
+    u = H.g / sqrt(prod(H.aspect))
+    interaction = ho_2brel_interaction(H.S, H.vtable, omm, OccupiedModeMap(new_addr))
```

Comment:
Why is `ho_2brel_interaction` accepting `OccupiedModeMap`s and why are they passed? It is only supposed to calculate the matrix element of a single particle excitation, so it will need to know `p_i` and `p_j`. If you make that change then it looks like this will work for multiple particle as well.